// @openshift-console/dynamic-plugin-sdk-webpack/lib/api/common-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrometheusEndpoint = exports.Operator = void 0;
var Operator;
(function (Operator) {
    Operator["Exists"] = "Exists";
    Operator["DoesNotExist"] = "DoesNotExist";
    Operator["In"] = "In";
    Operator["NotIn"] = "NotIn";
    Operator["Equals"] = "Equals";
    Operator["NotEqual"] = "NotEqual";
    Operator["GreaterThan"] = "GreaterThan";
    Operator["LessThan"] = "LessThan";
    Operator["NotEquals"] = "NotEquals";
})(Operator = exports.Operator || (exports.Operator = {}));
var BadgeType;
(function (BadgeType) {
    BadgeType["DEV"] = "Dev Preview";
    BadgeType["TECH"] = "Tech Preview";
})(BadgeType || (BadgeType = {}));
var PrometheusEndpoint;
(function (PrometheusEndpoint) {
    PrometheusEndpoint["LABEL"] = "api/v1/label";
    PrometheusEndpoint["QUERY"] = "api/v1/query";
    PrometheusEndpoint["QUERY_RANGE"] = "api/v1/query_range";
    PrometheusEndpoint["RULES"] = "api/v1/rules";
    PrometheusEndpoint["TARGETS"] = "api/v1/targets";
})(PrometheusEndpoint = exports.PrometheusEndpoint || (exports.PrometheusEndpoint = {}));
