// @openshift-console/dynamic-plugin-sdk-internal/lib/api/internal-api.js
export const ActivityItem = require('@console/shared/src/components/dashboard/activity-card/ActivityItem')
    .default;
export const ActivityBody = require('@console/shared/src/components/dashboard/activity-card/ActivityBody')
    .default;
export const RecentEventsBody = require('@console/shared/src/components/dashboard/activity-card/ActivityBody')
    .RecentEventsBody;
export const OngoingActivityBody = require('@console/shared/src/components/dashboard/activity-card/ActivityBody')
    .OngoingActivityBody;
export const AlertsBody = require('@console/shared/src/components/dashboard/status-card/AlertsBody')
    .default;
export const AlertItem = require('@console/shared/src/components/dashboard/status-card/AlertItem')
    .default;
export const HealthItem = require('@console/shared/src/components/dashboard/status-card/HealthItem')
    .default;
export const HealthBody = require('@console/shared/src/components/dashboard/status-card/HealthBody')
    .default;
export const ResourceInventoryItem = require('@console/shared/src/components/dashboard/inventory-card/InventoryItem')
    .ResourceInventoryItem;
export const DetailsBody = require('@console/shared/src/components/dashboard/details-card/DetailsBody')
    .default;
export const UtilizationItem = require('@console/shared/src/components/dashboard/utilization-card/UtilizationItem')
    .default;
export const UtilizationBody = require('@console/shared/src/components/dashboard/utilization-card/UtilizationBody')
    .default;
export const UtilizationDurationDropdown = require('@console/shared/src/components/dashboard/utilization-card/UtilizationDurationDropdown')
    .UtilizationDurationDropdown;
export const VirtualizedGrid = require('@console/shared/src/components/virtualized-grid/VirtualizedGrid')
    .default;
export const LazyActionMenu = require('@console/shared/src/components/actions/LazyActionMenu')
    .default;
export const QuickStartsLoader = require('@console/app/src/components/quick-starts/loader/QuickStartsLoader')
    .default;
export const useUtilizationDuration = require('@console/shared/src/hooks/useUtilizationDuration')
    .useUtilizationDuration;
export const useActiveNamespace = require('@console/shared/src/hooks/useActiveNamespace')
    .useActiveNamespace;
export const ServicesList = require('@console/internal/components/service').ServicesList;
export const useDashboardResources = require('@console/shared/src/hooks/useDashboardResources')
    .useDashboardResources;
export const useUserSettings = require('@console/shared/src/hooks/useUserSettings')
    .useUserSettings;
