// @openshift-console/dynamic-plugin-sdk-internal/lib/extensions/console-types.js
var InventoryStatusGroup;
(function (InventoryStatusGroup) {
    InventoryStatusGroup["WARN"] = "WARN";
    InventoryStatusGroup["ERROR"] = "ERROR";
    InventoryStatusGroup["PROGRESS"] = "PROGRESS";
    InventoryStatusGroup["NOT_MAPPED"] = "NOT_MAPPED";
    InventoryStatusGroup["UNKNOWN"] = "UNKNOWN";
})(InventoryStatusGroup || (InventoryStatusGroup = {}));
export var HealthState;
(function (HealthState) {
    HealthState["OK"] = "OK";
    HealthState["ERROR"] = "ERROR";
    HealthState["WARNING"] = "WARNING";
    HealthState["LOADING"] = "LOADING";
    HealthState["UNKNOWN"] = "UNKNOWN";
    HealthState["UPDATING"] = "UPDATING";
    HealthState["PROGRESS"] = "PROGRESS";
    HealthState["UPGRADABLE"] = "UPGRADABLE";
    HealthState["NOT_AVAILABLE"] = "NOT_AVAILABLE";
})(HealthState || (HealthState = {}));
export var GridPosition;
(function (GridPosition) {
    GridPosition["MAIN"] = "MAIN";
    GridPosition["LEFT"] = "LEFT";
    GridPosition["RIGHT"] = "RIGHT";
})(GridPosition || (GridPosition = {}));
export var LIMIT_STATE;
(function (LIMIT_STATE) {
    LIMIT_STATE["ERROR"] = "ERROR";
    LIMIT_STATE["WARN"] = "WARN";
    LIMIT_STATE["OK"] = "OK";
})(LIMIT_STATE || (LIMIT_STATE = {}));
