// @openshift-console/dynamic-plugin-sdk-webpack/lib/extensions/navigation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isNavExtension = exports.isNavItemOrSeparator = exports.isNavItem = exports.isNavSection = exports.isSeparator = exports.isResourceClusterNavItem = exports.isResourceNSNavItem = exports.isHrefNavItem = void 0;
// Type guards
const isHrefNavItem = (e) => e.type === 'console.navigation/href';
exports.isHrefNavItem = isHrefNavItem;
const isResourceNSNavItem = (e) => e.type === 'console.navigation/resource-ns';
exports.isResourceNSNavItem = isResourceNSNavItem;
const isResourceClusterNavItem = (e) => e.type === 'console.navigation/resource-cluster';
exports.isResourceClusterNavItem = isResourceClusterNavItem;
const isSeparator = (e) => e.type === 'console.navigation/separator';
exports.isSeparator = isSeparator;
const isNavSection = (e) => e.type === 'console.navigation/section';
exports.isNavSection = isNavSection;
const isNavItem = (e) => exports.isHrefNavItem(e) || exports.isResourceNSNavItem(e) || exports.isResourceClusterNavItem(e);
exports.isNavItem = isNavItem;
const isNavItemOrSeparator = (e) => exports.isNavItem(e) || exports.isSeparator(e);
exports.isNavItemOrSeparator = isNavItemOrSeparator;
const isNavExtension = (e) => exports.isNavItem(e) || exports.isSeparator(e) || exports.isNavSection(e);
exports.isNavExtension = isNavExtension;
