// @openshift-console/dynamic-plugin-sdk-webpack/lib/extensions/topology.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isTopologyRelationshipProvider = exports.isTopologyDecoratorProvider = exports.isTopologyDisplayFilters = exports.isTopologyDataModelFactory = exports.isTopologyCreateConnector = exports.isTopologyComponentFactory = void 0;
// Type Guards
const isTopologyComponentFactory = (e) => e.type === 'console.topology/component/factory';
exports.isTopologyComponentFactory = isTopologyComponentFactory;
const isTopologyCreateConnector = (e) => e.type === 'console.topology/create/connector';
exports.isTopologyCreateConnector = isTopologyCreateConnector;
const isTopologyDataModelFactory = (e) => e.type === 'console.topology/data/factory';
exports.isTopologyDataModelFactory = isTopologyDataModelFactory;
const isTopologyDisplayFilters = (e) => e.type === 'console.topology/display/filters';
exports.isTopologyDisplayFilters = isTopologyDisplayFilters;
const isTopologyDecoratorProvider = (e) => e.type === 'console.topology/decorator/provider';
exports.isTopologyDecoratorProvider = isTopologyDecoratorProvider;
const isTopologyRelationshipProvider = (e) => e.type === 'console.topology/relationship/provider';
exports.isTopologyRelationshipProvider = isTopologyRelationshipProvider;
