// @openshift-console/dynamic-plugin-sdk-webpack/lib/extensions/topology-details.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isNetworkAdapter = exports.isBuildAdapter = exports.isPodAdapter = exports.isDetailsResourceAlert = exports.isDetailsResourceLink = exports.isDetailsTabSection = exports.isDetailsTab = void 0;
// Type guards
const isDetailsTab = (e) => {
    return e.type === 'console.topology/details/tab';
};
exports.isDetailsTab = isDetailsTab;
const isDetailsTabSection = (e) => {
    return e.type === 'console.topology/details/tab-section';
};
exports.isDetailsTabSection = isDetailsTabSection;
const isDetailsResourceLink = (e) => {
    return e.type === 'console.topology/details/resource-link';
};
exports.isDetailsResourceLink = isDetailsResourceLink;
const isDetailsResourceAlert = (e) => {
    return e.type === 'console.topology/details/resource-alert';
};
exports.isDetailsResourceAlert = isDetailsResourceAlert;
const isPodAdapter = (e) => {
    return e.type === 'console.topology/adapter/pod';
};
exports.isPodAdapter = isPodAdapter;
const isBuildAdapter = (e) => {
    return e.type === 'console.topology/adapter/build';
};
exports.isBuildAdapter = isBuildAdapter;
const isNetworkAdapter = (e) => {
    return e.type === 'console.topology/adapter/network';
};
exports.isNetworkAdapter = isNetworkAdapter;
